src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1453,42 +1453,110 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
+                            intOrVoidEntryPoints.Add(candidate);
+                        }
+                        perCandidateBag.Free();
                     }
+                }
+
+                var viableEntryPoints = ArrayBuilder<MethodSymbol>.GetInstance();
 
+                foreach (var candidate in intOrVoidEntryPoints)
+                {
+                    if (candidate.IsAsync)
+                    {
+                        diagnostics.Add(ErrorCode.ERR_NonTaskMainCantBeAsync, candidate.Locations.First(), candidate);
+                        continue;
+                    }
                     viableEntryPoints.Add(candidate);
                 }
 
-                if ((object)mainType == null || viableEntryPoints.Count == 0)
+                if (viableEntryPoints.Count == 0)
+                {
+                    foreach (var (IsValid, Candidate, SpecificDiagnostics) in taskEntryPoints)
+                    {
+                        if (!CheckValid(Candidate, IsValid, SpecificDiagnostics))
+                        {
+                            continue;
+                        }
+
+                        // PROTOTYPE(async-main): Get the diagnostic to point to a smaller syntax piece.
+                        if (CheckFeatureAvailability(Candidate.GetNonNullSyntaxNode(), MessageID.IDS_FeatureAsyncMain, diagnostics))
+                        {
+                            diagnostics.AddRange(SpecificDiagnostics);
+                            viableEntryPoints.Add(Candidate);
+                        }
+                    }
+                }
+
+                foreach (var (_, _, SpecificDiagnostics) in taskEntryPoints)
                 {
-                    diagnostics.AddRange(warnings);
+                    SpecificDiagnostics.Free();
                 }
 
-                warnings.Free();
+                if (viableEntryPoints.Count == 0)
+                {
+                    diagnostics.AddRange(noMainFoundDiagnostics);
+                } 
+                else if ((object)mainType == null)
+                {
+                    // Turns error diagnostics into warnings. The reason that Error diagnostics can end up in `noMainFoundDiagnostics` is when 
+                    // HasEntryPointSignature yields some Error Diagnostics when people implement Task or Task<T> incorrectly.
+                    //
+                    // We can't add those Errors to the general diagnostics bag because it would break previously-working programs.
+                    // The fact that these warnings are not added when csc is invoked with /main is possibly a bug, and is tracked at
+                    // https://github.com/dotnet/roslyn/issues/18964
+                    foreach (var diagnostic in noMainFoundDiagnostics.AsEnumerable())
+                    {
+                        if (diagnostic.Severity == DiagnosticSeverity.Error) {
+                            diagnostics.Add(diagnostic.WithSeverity(DiagnosticSeverity.Warning));
```

Comment:
>diagnostics.Add(diagnostic.WithSeverity(DiagnosticSeverity.Warning)); [](start = 28, length = 69)

I don't think changing severity on the diagnostics is the right thing to do. Also, why, for example, would we want to report an obsolete diagnostics associated with GetAwaiter for an entry point that isn't going to be used? Please add a test for this scenario. I still think that we should be reporting only specific warnings, which are reported in CheckValid.